src/main/java/minijava/ir/optimize/BaseOptimizer.java

```diff
@@ -4,33 +4,42 @@
 import firm.Graph;
 import firm.nodes.Node;
 import firm.nodes.NodeVisitor;
+import java.util.Collection;
+import minijava.ir.utils.FirmUtils;
 
 public abstract class BaseOptimizer extends NodeVisitor.Default implements Optimizer {
 
   protected Graph graph;
   protected boolean hasChanged;
 
   /**
-   * Implements the work-list algorithm.
+   * Implements the work-list algorithm, populating the initial worklist by passing .
    *
    * <p>Nodes in the work-list are processed by calling {@link Node#accept(NodeVisitor)
    * node.accept(this)}. In order for the data-flow analyses to be successful, implementations of
    * {@link BaseOptimizer} <b>must</b> set the protected {@code hasChanged} field to true in the
    * overwritten {@code visit(Node)} methods, if the outputs of the node's transfer function differ
    * from the outputs of the previous visiting.
    */
-  protected void fixedPointIteration() {
-    Worklist worklist = Worklist.fillTopological(graph);
-    while (!worklist.isEmpty()) {
-      Node n = worklist.removeFirst();
-      hasChanged = false;
-      n.accept(this);
-      if (hasChanged) {
-        for (BackEdges.Edge e : BackEdges.getOuts(n)) {
-          worklist.addFirst(e.node);
-        }
-      }
-    }
+  protected boolean fixedPointIteration(Collection<Node> initialWorklist) {
+    return FirmUtils.withBackEdges(
+        graph,
+        () -> {
+          boolean hasChangedAtAll = false;
+          Worklist worklist = new Worklist(initialWorklist);
+          while (!worklist.isEmpty()) {
+            Node n = worklist.dequeue();
+            hasChanged = false;
+            n.accept(this);
+            if (hasChanged) {
+              for (BackEdges.Edge e : BackEdges.getOuts(n)) {
+                worklist.enqueue(e.node);
+              }
+              hasChangedAtAll = true;
```

Comment:
The names are not particularly expressive... Maybe `currentNodeInducedChange` and `graphHasChanged`?